src/zenml/zen_server/template_execution/utils.py

```diff
@@ -50,26 +51,85 @@
 from zenml.zen_server.template_execution.runner_entrypoint_configuration import (
     RunnerEntrypointConfiguration,
 )
-from zenml.zen_server.utils import server_config, workload_manager, zen_store
+from zenml.zen_server.utils import (
+    run_template_executor,
+    server_config,
+    workload_manager,
+    zen_store,
+)
 
 logger = get_logger(__name__)
 
 RUNNER_IMAGE_REPOSITORY = "zenml-runner"
 
 
+class BoundedThreadPoolExecutor:
+    """Thread pool executor with a queue size limit."""
+
+    def __init__(
+        self, max_queue_size: int, max_workers: int, **kwargs: Any
+    ) -> None:
+        """Initialize the executor.
+
+        Args:
+            max_queue_size: The maximum number of tasks in the queue.
+            max_workers: The maximum number of workers.
+            **kwargs: Arguments to pass to the thread pool executor.
+        """
+        self._executor = ThreadPoolExecutor(max_workers=max_workers, **kwargs)
+        self._semaphore = threading.BoundedSemaphore(
+            value=max_queue_size + max_workers
+        )
+
+    def submit(
+        self, fn: Callable[..., Any], *args: Any, **kwargs: Any
+    ) -> Future[Any]:
+        """Submit a task to the executor.
+
+        Args:
+            fn: The function to execute.
+            *args: The arguments to pass to the function.
+            **kwargs: The keyword arguments to pass to the function.
+
+        Raises:
+            Exception: If the task submission fails.
+
+        Returns:
+            The future of the task.
+        """
+        self._semaphore.acquire()
+        try:
+            future = self._executor.submit(fn, *args, **kwargs)
+        except Exception:
+            self._semaphore.release()
+            raise
+        else:
+            future.add_done_callback(lambda _: self._semaphore.release())
+            return future
+
+    def shutdown(self, **kwargs: Any) -> None:
+        """Shutdown the executor.
+
+        Args:
+            **kwargs: Keyword arguments to pass to the shutdown method of the
+                executor.
+        """
+        self._executor.shutdown(**kwargs)
```

Comment:
This isn't called anywhere. Are you missing the cleanup/deinitialization part ?

Reply:
Added the call in the fastapi shutdown event handler